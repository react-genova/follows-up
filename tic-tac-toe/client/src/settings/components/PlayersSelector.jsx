import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import PlayersSelectorStyled from './PlayersSelector.styled';

const PlayersSelector = ({ onClick, children }) => {
    const [hover, setHover] = useState(false);
    const onMouseEnter = useMemo(() => () => setHover(true), []);
    const onMouseLeave = useMemo(() => () => setHover(false), []);
    return (
        <PlayersSelectorStyled
            onClick={onClick}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
        >
            {children(hover ? '#E0E0E0' : '#909090')}
        </PlayersSelectorStyled>
    );
};

PlayersSelector.propTypes = {
    onClick: PropTypes.func,
    children: PropTypes.func.isRequired,
};

PlayersSelector.defaultProps = {
    onClick: undefined,
};

export default PlayersSelector;
