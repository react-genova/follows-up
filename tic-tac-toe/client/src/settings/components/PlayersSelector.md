### Han Solo

```jsx
const { Fragment } = require('react');
<div style={{ backgroundColor: '#303030', width: '35em' }}>
    <PlayersSelector>
    {
        (color) => (
            <Fragment>
                <Human color={color} />
                <Versus color={color} />
                <Machine color={color} />
            </Fragment>
        )
    }
    </PlayersSelector>
</div>
```