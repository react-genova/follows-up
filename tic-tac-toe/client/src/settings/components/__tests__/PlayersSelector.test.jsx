import React from 'react';
import { fireEvent, render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import PlayersSelector from '../PlayersSelector';

describe('PlayersSelector component', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        const { container } = render(<PlayersSelector>{() => null}</PlayersSelector>);
        expect(container.firstChild).toMatchSnapshot();
    });
// '#E0E0E0' : '#909090'
    it('invokes child as function with #E0E0E0 on mount', () => {
        const CHILDREN = jest.fn();
        render(<PlayersSelector>{CHILDREN}</PlayersSelector>);
        expect(CHILDREN).toHaveBeenCalledTimes(1);
        expect(CHILDREN).toHaveBeenCalledWith('#909090');
    });

    it('invokes child as function with #E0E0E0 on mount', () => {
        const CHILDREN = jest.fn();
        const onClick = jest.fn();
        const { container } = render(<PlayersSelector onClick={onClick}>{CHILDREN}</PlayersSelector>);
        
        fireEvent.click(container.firstChild);
        expect(onClick).toHaveBeenCalledTimes(1);

        fireEvent.mouseEnter(container.firstChild);
        fireEvent.mouseLeave(container.firstChild);
        expect(CHILDREN).toHaveBeenCalledTimes(3);
        expect(CHILDREN).toHaveBeenCalledWith('#909090');
        expect(CHILDREN).toHaveBeenCalledWith('#E0E0E0');
        expect(CHILDREN).toHaveBeenLastCalledWith('#909090');
    });
});
