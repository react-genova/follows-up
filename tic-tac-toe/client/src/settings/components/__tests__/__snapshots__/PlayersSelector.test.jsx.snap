// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersSelector component renders without crashing 1`] = `
<div
  class="sc-bdVaJa iOktxA"
/>
`;
